README
NAME

Hash::SharedMem - efficient shared mutable hash

DESCRIPTION

This module provides a facility for efficiently sharing mutable data
between processes on one host.  Data is organised as a key/value store,
resembling a Perl hash.  The keys and values are restricted to octet
(Latin-1) strings.  Structured objects may be stored by serialising them
using a mechanism such as Sereal.

The data is represented in files that are mapped into each process's
memory space, which for interprocess communication amounts to the
processes sharing memory.  Processes are never blocked waiting for each
other.  The use of files means that there is some persistence, with the
data continuing to exist when there are no processes with it mapped.

The data structure is optimised for the case where all the data fits
into RAM.  This happens either via buffering of a disk-based filesystem,
or as the normal operation of a memory-backed filesystem, in either case
as long as there isn't much swapping activity.  If RAM isn't large enough,
such that the data has to reside mainly on disk and parts of it have to
be frequently reread from disk, speed will seriously suffer.  The data
structure exhibits poor locality of reference, and is not designed to
play nicely with filesystem block sizes.

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2014 PhotoBox Ltd

Copyright (C) 2014 Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.003; 2014-05-05

  * bugfix: on Perl 5.6, repair a dodgy workaround that could cause
    magical (e.g., tainted) arguments to lose their magic flags

  * bugfix: build custom ops in a way that satisfies an unnecessary
    assertion in debugging Perl builds

  * new function shash_idle() to prevent an idle handle keeping an
    obsolete data file around

  * when tidying, include a non-zero constant term in the computation
    of the new file size, to avoid cycling data files very rapidly when
    the shared hash contains little data

  * count events for profiling and debugging, counters accessed
    through new functions shash_tally_get(), shash_tally_zero(), and
    shash_tally_gzero()

  * make the constant shash_referential_handle also available as a class
    method on Hash::SharedMem::Handle

  * when attempting to unlink a file, other than for cleanup of an
    operation that died, if unlinking fails for reasons other than ENOENT
    or EBUSY, report the error, rather than silently ignoring it

  * cooperate in the tainting system, mainly imitating the taint behaviour
    of regular file handles

  * support using the module in multiple threads simultaneously,
    including, on Perl 5.8.9 and later, duplicating handles for thread
    spawning

  * cope on Cygwin, which offers a fraudulent openat(2) et al

  * avoid a C99 declaration-after-statement construction

  * avoid using "NULL" for null function pointers, for C compiler
    portability

  * be cleverer about the asymmetry of key ranges in the B-tree structure,
    saving about one key comparison per data operation

  * use cursors internally, so that the key has to be resolved only
    once for each combined read-and-write operation (shash_gset()
    and shash_cset())

  * when attempting a data file rollover, if the attempt fails due to
    a conflicting update from another process, unmap the abortive data
    file before attempting to unlink it, so that the unlink will succeed
    on OSes that need a mapped file to remain linked

  * in documentation, comment on the inefficiency that occurs when the
    shared hash doesn't fit into RAM

  * in documentation, comment on the Unicode bug in the treatment of
    filenames

  * don't uselessly attempt to hook into B::Deparse on Perls where custom
    ops aren't registered and so the hook can't work

  * don't uselessly attempt to hook into B::Deparse for the
    shash_referential_handle constant which isn't a custom op type

  * test concurrency with real concurrent racing processes

  * test that a forked handle works correctly

  * include magic(5) file, describing the identifying part of the format
    of shared hash files, in the distribution

  * manage destruction of shared hash handles through magic rather than
    a DESTROY method, for a small speedup

  * restructure the internal representation of directory references,
    for a small simplification of uses of the references

  * in test suite, extend to shash_tidy() some patterns of tests that
    cover most operators and which were missed when shash_tidy() was
    added to the module

  * slight refactoring of C code

  * on systems that lack mmap(2), detect the problem cleanly and report
    it in a way that CPAN Testers understand

  * in design document and code, use the more standard term "fanout"
    instead of "splay"

  * take todo notes out of code comments

version 0.002; 2014-03-25

  * new function shash_tidy() to make data file rollover happen at a
    convenient time

  * in many error messages, state more specifically and consistently
    what kind of action has failed

  * in documentation, consistently describe the content of a shared hash
    as "content" rather than "contents"

version 0.001; 2014-03-23

  * bugfix: avoid overflowing a directory entry buffer on OSes where
    struct dirent is not a suitable size for a buffer

  * bugfix: avoid overflowing a filename buffer when time exceeds the
    32-bit range (a year 2038 problem) or PIDs are larger than 32 bits

  * ensure reliable behaviour across fork(2), by not holding on to
    allocated data file space between write operations

  * create all data files with the same permission bits as the master
    file, so that the permissions set when the shash was created stick,
    and the umask at the time of shash operations doesn't matter

  * attempt to give all data files the same group and owner as the master
    file, so that permissions behave as consistently as possible

  * new exported constant shash_referential_handle to indicate whether
    shared hash handles constitute first-class references to the
    underlying files

  * detect non-octet string values for key and value parameters reliably
    and early, consistently signalling the error

  * detect non-string values for filename parameters, signalling the error

  * process get magic on each parameter exactly once per hash operation
    (though users of the tied interface are at the mercy of the tying
    infrastructure, which doesn't have such clean behaviour)

  * when attempting to clean a shash directory, if listing the directory
    fails, report the error, rather than silently abandoning the cleaning
    attempt

  * when attempting to generate a new data file, if the size is so large
    as to cast to a negative off_t value, detect it early and report
    EFBIG, rather than going ahead to create the file and detect the
    error at ftruncate(2) time

  * if ftruncate(2) reports that a requested file size is too big using
    errno EINVAL, report it as the more enlightening EFBIG

  * when attempting to iterate a shash directory, if opening the directory
    fails, detect it and report the real error, rather than going ahead
    to fdopendir(3) and detecting its EBADF

  * when reporting an error including the name of a shash, to match
    user expectations use the characters of the filename scalar that

Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4205
require 5.014;

    unless (eval "use Module::Build::Compat 0.02; 1" ) {
      print "This module requires Module::Build to install itself.\n";

      require ExtUtils::MakeMaker;
      my $yn = ExtUtils::MakeMaker::prompt
	('  Install Module::Build now from CPAN?', 'y');

      unless ($yn =~ /^y/i) {
	die " *** Cannot install without Module::Build.  Exiting ...\n";
      }

      require Cwd;
      require File::Spec;
      require CPAN;

      # Save this 'cause CPAN will chdir all over the place.
      my $cwd = Cwd::cwd();

      CPAN::Shell->install('Module::Build::Compat');
      CPAN::Shell->expand("Module", "Module::Build::Compat")->uptodate
	or die "Couldn't install Module::Build, giving up.\n";

      chdir $cwd or die "Cannot chdir() back to $cwd: $!";
    }
    eval "use Module::Build::Compat 0.02; 1" or die $@;
    
    Module::Build::Compat->run_build_pl(args => \@ARGV);
    my $build_script = 'Build';
    $build_script .= '.com' if $^O eq 'VMS';
    exit(0) unless(-e $build_script); # cpantesters convention
    require Module::Build;
    Module::Build::Compat->write_makefile(build_class => 'Module::Build');


lib/Hash/SharedMem/Handle.pm
=head1 NAME

Hash::SharedMem::Handle - handle for efficient shared mutable hash

=head1 SYNOPSIS

	use Hash::SharedMem::Handle;

	if(Hash::SharedMem::Handle->referential_handle) { ...

	$shash = Hash::SharedMem::Handle->open($filename, "rwc");

	if($shash->is_readable) { ...
	if($shash->is_writable) { ...
	$mode = $shash->mode;

	if($shash->getd($key)) { ...
	$value = $shash->get($key);
	$shash->set($key, $newvalue);
	$oldvalue = $shash->gset($key, $newvalue);
	if($shash->cset($key, $chkvalue, $newvalue)) { ...

	$snap_shash = $shash->snapshot;
	if($shash->is_snapshot) { ...

	$shash->idle;
	$shash->tidy;

	$tally = $shash->tally_get;
	$shash->tally_zero;
	$tally = $shash->tally_gzero;

	tie %shash, "Hash::SharedMem::Handle", $shash;
	tie %shash, "Hash::SharedMem::Handle", $filename, "rwc";

	$shash = tied(%shash);
	if(exists($shash{$key})) { ...
	$value = $shash{$key};
	$shash{$key} = $newvalue;
	$oldvalue = delete($shash{$key});

=head1 DESCRIPTION

An object of this class is a handle referring to a memory-mapped shared
hash object of the kind described in L<Hash::SharedMem>.  It can be
passed to the functions of that module, or the same operations can be
performed by calling the methods described below.  Uses of the function
and method interfaces may be intermixed arbitrarily; they are completely
equivalent in function.  They are not equivalent in performance, however,
with the method interface being somewhat slower.

This class also supplies a tied-hash interface to shared hashes.  The tied
interface is much slower than the function and method interfaces.
The behaviour of a tied hash more resembles the function and method
interfaces to shared hashes than it resembles the syntactically-similar
use of ordinary Perl hashes.  Using a non-string as a key will result
in an exception, rather than stringification of the key.  Using a
string containing a non-octet codepoint as a key will also result in an
exception, rather than merely referring to an absent hash element.

=cut

package Hash::SharedMem::Handle;

{ use 5.006; }
use warnings;
use strict;

use Hash::SharedMem ();

our $VERSION = "0.003";

=head1 CLASS METHODS

=over

=item Hash::SharedMem::Handle->referential_handle

Returns a truth value indicating whether each shared hash handle
contains a first-class reference to the shared hash to which it refers.
See L<Hash::SharedMem/"Filesystem referential integrity"> for discussion
of the significance of this.

=back

=head1 CONSTRUCTOR

=over

=item Hash::SharedMem::Handle->open(FILENAME, MODE)

Opens and return a handle referring to a shared hash object,
or C<die>s if the shared hash can't be opened as specified.
See L<Hash::SharedMem/shash_open> for details.

=back

=head1 METHODS

=over

=item $shash->is_readable

=item $shash->is_writable

=item $shash->mode

=item $shash->getd(KEY)

=item $shash->get(KEY)

=item $shash->set(KEY, NEWVALUE)

=item $shash->gset(KEY, NEWVALUE)

=item $shash->cset(KEY, CHKVALUE, NEWVALUE)

=item $shash->snapshot

=item $shash->is_snapshot

=item $shash->idle

=item $shash->tidy

=item $shash->tally_get

=item $shash->tally_zero

=item $shash->tally_gzero

These methods are each equivalent to the corresponding
"C<shash_>"-prefixed function in L<Hash::SharedMem>.  See that document
for details.

=back

=head1 TIE CONSTRUCTORS

=over

=item tie(VARIABLE, "Hash::SharedMem::Handle", SHASH)

I<VARIABLE> must be a hash variable, and I<SHASH> must be a handle
referring to a shared hash object.  The call binds the variable to the
shared hash, so that the variable provides a view of the shared hash
that resembles an ordinary Perl hash.  The shared hash handle is returned.

=item tie(VARIABLE, "Hash::SharedMem::Handle", FILENAME, MODE)

I<VARIABLE> must be a hash variable.  The call opens a handle referring
to a shared hash object, as described in L<Hash::SharedMem/shash_open>,
and binds the variable to the shared hash, so that the variable provides a
view of the shared hash that resembles an ordinary Perl hash.  The shared
hash handle is returned.

=back

=head1 TIED OPERATORS

For all of these operators, the key of interest (I<KEY> parameter)
and values can each be any octet (Latin-1) string.  Strings containing
non-octets (Unicode characters above U+FF) and items other than strings
cannot be used as keys or values.  If a dualvar (scalar with independent
string and numeric values) is supplied, only its string value will
be used.

=over

=item tied(%SHASH)

Returns the handle via which I<%SHASH> is bound to the shared hash.
This is a shared hash handle that can be used by calling the methods
described above or by passing it to the functions of L<Hash::SharedMem>.

=item exists($SHASH{KEY})

Returns a truth value indicating whether the specified key is currently
present in the shared hash.

=item $SHASH{KEY}

Returns the value currently referenced by the specified key in the shared
hash, or C<undef> if the key is absent.

=item $SHASH{KEY} = NEWVALUE

Modifies the shared hash so that the specified key henceforth references
the specified value.  The new value must be a string.

=item delete($SHASH{KEY})

Modifies the shared hash so that the specified key is henceforth absent,
and returns the value that the key previously referenced, or C<undef>
if the key was already absent.  This swap is performed atomically.

=item %SHASH = LIST

Setting the entire content of the shared hash (throwing away the previous
content) is not supported.

=item each(%SHASH)

=item keys(%SHASH)

=item values(%SHASH)

=item %SHASH

Iteration over, or enumeration of, the shared hash's content is not
supported.

=item scalar(%SHASH)

Checking whether the shared hash is occupied is not supported.

=back

=head1 BUGS

Due to details of the Perl implementation, this object-oriented interface
to the shared hash mechanism is somewhat slower than the function
interface, and the tied interface is much slower.  The functions in
L<Hash::SharedMem> are the recommended interface.

=head1 SEE ALSO

L<Hash::SharedMem>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2014 PhotoBox Ltd

Copyright (C) 2014 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


Build.PL
{ use 5.006; }
use warnings;
use strict;

use Module::Build;

Module::Build->subclass(code => q{
	unless(__PACKAGE__->can("cbuilder")) {
		*cbuilder = sub { $_[0]->_cbuilder or die "no C support" };
	}
	sub probed_info {
		my($self) = @_;
		my $feat_file = $self->localize_file_path("lib/Hash/features");
		my $probe_file = $self->localize_file_path(
				"lib/Hash/features.probe");
		unless($self->up_to_date($probe_file, $feat_file)) {
			require IO::File;
			require Data::Dumper;
			my $probe = do($probe_file) or
				die $@ || "can't read $probe_file: $!";
			my $feat_defs = $probe->($self);
			$self->add_to_cleanup($feat_file);
			my $feat_fh = IO::File->new($feat_file, "w")
				or die "can't write $feat_file: $!";
			local $Data::Dumper::Terse = 1;
			$feat_fh->printflush(Data::Dumper::Dumper($feat_defs))
				or die "can't write $feat_file: $!";
		}
		my $probed_info = do($feat_file) or
			die $@ || "can't read $feat_file: $!";
		return ($probed_info, $feat_file);
	}
	sub compile_c {
		my($self, $file, %args) = @_;
		my $cc0_h =
			$self->localize_file_path("lib/Hash/callchecker0.h");
		unless(-f $cc0_h) {
			my $content = eval {
				local $SIG{__DIE__};
				require Devel::CallChecker;
				Devel::CallChecker->VERSION(0.003);
				&Devel::CallChecker::callchecker0_h();
			} || "";
			$self->add_to_cleanup($cc0_h);
			require IO::File;
			my $fh = IO::File->new($cc0_h, "w") or die $!;
			$fh->printflush($content) or die $!;
			$fh->close or die $!;
		}
		unless($args{no_feature_defs}) {
			my($probed, $probed_file) = $self->probed_info;
			$args{defines} = {
				%{$args{defines} || {}},
				%{$probed->{defs}},
			};
			$obj_file = $self->cbuilder->object_file($file);
			unless($self->up_to_date($probed_file, $obj_file)) {
				unlink $obj_file;
			}
		}
		return $self->SUPER::compile_c($file, %args);
	}
	sub link_c {
		no strict "refs";
		my($self, $spec) = @_;
		my($probed, $probed_file) = $self->probed_info;
		unless($self->up_to_date($probed_file, $spec->{lib_file})) {
			unlink $spec->{lib_file};
		}
		my $cb = $self->cbuilder;
		my $cbclass = ref($cb);
		my $orig_cb_link = $cb->can("link");
		local *{"${cbclass}::link"} = sub {
			my($self, %args) = @_;
			if($args{module_name} eq "Hash::SharedMem") {
				my $cc_linkables = eval {
				    local $SIG{__DIE__};
				    require Devel::CallChecker;
				    Devel::CallChecker->VERSION(0.003);
				    [&Devel::CallChecker::callchecker_linkable];
				} || [];
				$args{objects} = [
					@{$args{objects}},
					@$cc_linkables,
				];
			}
			$args{extra_linker_flags} = [
				@{$args{extra_linker_flags} || []},
				@{$probed->{libs}},
			];
			@_ = ($self, %args);
			goto &$orig_cb_link;
		};
		$self->SUPER::link_c($spec);
	}
})->new(
	module_name => "Hash::SharedMem",
	license => "perl",
	configure_requires => {
		"Module::Build" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	build_requires => {
		"Data::Dumper" => 0,
		"Errno" => "1.00",
		"ExtUtils::CBuilder" => "0.15",
		"File::Temp" => "0.22",
		"IO::File" => 0,
		"Module::Build" => 0,
		"POSIX" => 0,
		"Scalar::String" => 0,
		"Test::More" => 0,
		"if" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"utf8" => 0,
		"warnings" => 0,
	},
	build_recommends => {
		"Devel::CallChecker" => "0.003",
	},
	requires => {
		"Exporter" => 0,
		"XSLoader" => 0,
		"parent" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	recommends => {
		"Devel::CallChecker" => "0.003",
	},
	dynamic_config => 0,
	meta_add => { distribution_type => "module" },
	create_makefile_pl => "passthrough",
	sign => 1,
)->create_build_script;

1;
